src/providers/ibmcloud/gen2.rs

```diff
@@ -0,0 +1,196 @@
+//! Metadata fetcher for IBMCloud (VPC Gen2) instances.
+//!
+//! This provider supports the "VPC Generation 2" infrastructure type
+//! on IBMCloud.
+//! It provides a config-drive as the only metadata source, whose layout
+//! is very similar to `cloud-init NoCloud` [datasource][nocloud], with
+//! a few variations:
+//!  - disk label is `cidata` (lowercase)
+//!  - filesystem is `iso9660`
+//!
+//! nocloud: https://cloudinit.readthedocs.io/en/latest/topics/datasources/nocloud.html
+
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{BufRead, BufReader, Read};
+use std::path::{Path, PathBuf};
+
+use openssh_keys::PublicKey;
+use slog_scope::{error, warn};
+use tempdir::TempDir;
+
+use crate::errors::*;
+use crate::network;
+use crate::providers::MetadataProvider;
+
+const CONFIG_DRIVE_LABEL: &str = "cidata";
+
+/// IBMCloud provider (VPC Gen2).
+#[derive(Debug)]
+pub struct G2Provider {
+    /// Path to the top directory of the mounted config-drive.
+    drive_path: PathBuf,
+    /// Temporary directory for own mountpoint (if any).
+    temp_dir: Option<TempDir>,
+}
+
+impl G2Provider {
+    /// Try to build a new provider client.
+    ///
+    /// This internally tries to mount (and own) the config-drive.
+    pub fn try_new() -> Result<Self> {
+        let target =
+            TempDir::new("afterburn").chain_err(|| "failed to create temporary directory")?;
+        crate::util::mount_ro(
+            &Path::new("/dev/disk/by-label/").join(CONFIG_DRIVE_LABEL),
+            target.path(),
+            "iso9660",
+            3,
+        )?;
+
+        let provider = Self {
+            drive_path: target.path().to_owned(),
+            temp_dir: Some(target),
+        };
+        Ok(provider)
```

Comment:
Minor: could squash this with previous statement?

Reply:
We can, but I don't in general as I prefer an additional line over an additional nesting level. Do you maybe know why your eyes picked this one specifically? How strong is your feeling on this?

Reply:
> How strong is your feeling on this?

Not very. :)  Agreed it comes down to personal preference.